src/core/dep-graph.ts

```diff
@@ -327,39 +308,60 @@ class DepGraphImpl implements types.DepGraphInternal {
     return node;
   }
 
-  private pathsFromNodeToRoot(nodeId: string): types.PkgInfo[][] {
+  private pathsFromNodeToRoot(
+    nodeId: string,
+    seenOnPath: Set<string>,
+  ): string[][] {
     const parentNodesIds = this.getNodeParentsNodeIds(nodeId);
-    if (parentNodesIds.length === 0) {
-      return [[this.getNodePkg(nodeId)]];
+
+    // No parents --> we're at root!
+    if (!parentNodesIds.length) {
+      return [[nodeId]];
     }
 
-    const allPaths: types.PkgInfo[][] = [];
-    parentNodesIds.map((id) => {
-      const out = this.pathsFromNodeToRoot(id).map((path) => {
-        return [this.getNodePkg(nodeId)].concat(path);
+    const allPaths: string[][] = [];
+    parentNodesIds
+      .filter((parentId) => !seenOnPath.has(parentId))
+      .map((parentId) => {
+        const pathsFromParent = this.pathsFromNodeToRoot(
+          parentId,
+          new Set(seenOnPath).add(nodeId), // Max size is O(log(n)) [tree depth] where n is the amount of nodes in the tree
+        );
+
+        for (const path of pathsFromParent) {
+          allPaths.push([nodeId].concat(path));
+        }
       });
-      for (const path of out) {
-        allPaths.push(path);
-      }
-    });
 
     return allPaths;
   }
 
-  private countNodePathsToRoot(nodeId: string): number {
+  private countNodePathsToRoot(
+    nodeId: string,
+    seenOnPath: Set<string>,
+  ): number {
     if (this._countNodePathsToRootCache.has(nodeId)) {
       return this._countNodePathsToRootCache.get(nodeId) || 0;
     }
 
     const parentNodesIds = this.getNodeParentsNodeIds(nodeId);
+
     if (parentNodesIds.length === 0) {
       this._countNodePathsToRootCache.set(nodeId, 1);
       return 1;
     }
 
-    const count = parentNodesIds.reduce((acc, parentNodeId) => {
-      return acc + this.countNodePathsToRoot(parentNodeId);
-    }, 0);
+    const count = parentNodesIds
```

Comment:
can we do this without `filter` and `reduce`? It creates a lot of GC (throw away arrays and function closures) for a frequently hit method whose performance I'm really worried about?